### PR TITLE
Improved Cast Diagnostics when Literals are involved

### LINKED ISSUE #53822: [SR-11421] Checked Cast Diagnostics Should Be More Specific When Literals Are Involved
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-11421      |
   |Radar             | None         |
   |Original Reporter | @CodaFi      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, DiagnosticsQoI, StarterBug        |
   |Assignee          | mohit.athwani (JIRA)      |
   |Priority          | Medium      |

   md5: 54d2d8b997d41666f9024642eb576483

  </details>

**cloned to**:  
* [SR-12093](https://bugs.swift.org/browse/SR-12093) Checked Casts With Literals Are Deeply Unintuitive  

**Issue Description:**

Checked casts on literals can have surprising effects at face value

```swift
let character = "A" as Character
let bad = "A" as? Character
```

The former is a character literal and is statically coerced. The latter is a `String` literal that is being dynamically cast to a `Character`. This cast will always fail, and we note as much - this also works with `as!` force casts.

```swift
warning: cast from 'String' to unrelated type 'Character' always fails
```

We could definitely do a better job phrasing this warning to let the user know what went wrong and can even offer a fixit that changes the dynamic cast to a static cast if it's going to succeed.

### PR DESCRIPTION
**Pull request by Rajveer100**
Fixes #53822

Emphasised the diagnostic involving conditional downcast to coercion making it more specific adding some contrast between casting (`as?`) and coercing (`as`).

### CODE REVIEW

**Comment by Rajveer100** [comment]
@AnthonyLatsis Request your review.

**Review by AnthonyLatsis — COMMENTED** [review]
Looking good so far!

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by Rajveer100** [comment]
Please let me know more issues to solve! 

While I keep surfing and searching, often there are issues which are not marked as good-first-issues but I assume I can at least try some viable ones!

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [inline include/swift/AST/DiagnosticsSema.def:1187]: Small nitpick, through out all diagnostics we consistently use "inferred as"(we can find a few examples if search for "inferred as" filtering *.def files)
```suggestion
        "conditional downcast from literal inferred as %0 to unrelated type %1 always fails; "
```
So maybe we could go with that here as well? @AnthonyLatsis WDYT?

**Comment by LucianoPAlmeida** [comment]
We have to update the test cases for this new message as well =]

**Comment by Rajveer100** [comment]
Whoops! My bad!

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def:1187]: Yeah, either one is fine to me as well =]

**Comment by Rajveer100** [comment]
@swift-ci please smoke test macOS

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test macOS platform

**Comment by Rajveer100** [comment]
Is it because of access rights that the bot commands from my end didn't work?

File: test/expr/cast/literals_downcast.swift
Comment (AnthonyLatsis) [inline test/expr/cast/literals_downcast.swift:5]: Can you add a test with `as!` too? It’s not necessarily a *conditional* downcast, so I think we should remove that elaboration from the message. It’s not necessarily even a downcast for that matter, but that’s not that big of a deal.

**Review by AnthonyLatsis — COMMENTED** [review]
Yes, the bot won’t trigger without write access.

File: test/expr/cast/literals_downcast.swift
Comment (Rajveer100) [reply test/expr/cast/literals_downcast.swift:5]: Sure!

Comment (Rajveer100) [reply test/expr/cast/literals_downcast.swift:5]: Could you tell me the exact `expected-warning` that must be added for the `as!` case?

Comment (AnthonyLatsis) [reply test/expr/cast/literals_downcast.swift:5]: Have you tried running the test? It should print unexpected or incorrect warnings.

Comment (Rajveer100) [reply test/expr/cast/literals_downcast.swift:6]: I have made the changes.

**Comment by Rajveer100** [comment]
Any other additions/changes that's required?

File: test/expr/cast/literals_downcast.swift
Comment (AnthonyLatsis) [reply test/expr/cast/literals_downcast.swift:6]: As I mentioned earlier, we use the same error message for `as?` and `as!`, but the message says "conditional downcast". This is a good opportunity to get rid of this inaccuracy if you’re interested.

File: include/swift/AST/DiagnosticsSema.def
Comment (AnthonyLatsis) [reply include/swift/AST/DiagnosticsSema.def:1187]: Please wrap this to 80 chars too.

**Comment by Rajveer100** [comment]
@AnthonyLatsis I have made the changes...

**Review by AnthonyLatsis — APPROVED** [review]
Very nice, thank you! The final word belongs to the code owners.

One last nit from me: please add a "Sema" tag to the commit message.

**Comment by Rajveer100** [comment]
Thanks! I have done that too!

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Pull request merged**
